File: courseScheduleII/courseScheduleII.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

class Solution {
public:
    vector<int> findOrder(int numCourses, vector<pair<int, int>>& prerequisites) {
        vector<int> res;
        if (numCourses == 0)
            return res;
        vector<vector<int> > m(numCourses, vector<int>());
        vector<int> indegree(numCourses, 0);
        for (auto& p : prerequisites) {
            m[p.second].push_back(p.first);
            indegree[p.first]++;
        }
        queue<int> q;
        for (int i = 0; i < numCourses; i++) {
            if (indegree[i] == 0)
                q.push(i);
        }
        int count = 0;
        while (!q.empty()) {
            int node = q.front();
            q.pop();
            count++;
            res.push_back(node);
            for (int neighbor : m[node]) {
                if (--indegree[neighbor] == 0)
                    q.push(neighbor);
            }
        }
        if (count != numCourses)
            res.clear();
        return res;
    }
};

int main()
{
  Solution s;
  vector<pair<int, int> > prerequisites;
  prerequisites.push_back(make_pair(1,0));
  prerequisites.push_back(make_pair(2,0));
  prerequisites.push_back(make_pair(3,1));
  prerequisites.push_back(make_pair(3,2));
  vector<int> res = s.findOrder(4, prerequisites);
  for (int i : res)
    cout << i << " ";
  cout << endl;
  return 0;
}

File: 210.courseScheduleII/courseScheduleII.java
class Solution {
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        Map<Integer, Set<Integer>> graph = new HashMap<>();
        for (int[] pre : prerequisites) {
            if (!graph.containsKey(pre[1])) {
                graph.put(pre[1], new HashSet<>());
            }
            graph.get(pre[1]).add(pre[0]);
        }
        int[] indegree = new int[numCourses];
        for (Set<Integer> s : graph.values()) {
            for (Integer i : s) indegree[i]++;
        }
        Queue<Integer> q = new LinkedList<>();
        for (int i = 0; i < indegree.length; i++) {
            if (indegree[i] == 0) q.offer(i);
        }
        int visitedNum = 0;
        int[] res = new int[numCourses];
        while (!q.isEmpty()) {
            Integer node = q.poll();
            res[visitedNum++] = node;
            Set<Integer> neighbors = graph.containsKey(node) ? graph.get(node) : new HashSet<>();
            for (Integer neighbor : neighbors) {
                if (--indegree[neighbor] == 0) {
                    q.offer(neighbor);
                }
            }
        }
        if (visitedNum == numCourses) return res;
        else return new int[0];
    }
}
